src/ApplicationSampler.cpp

```diff
@@ -407,10 +409,18 @@ namespace geopm
                                                                   core_idx);
                 GEOPM_DEBUG_ASSERT(inactive_cpu.size() != 0,
                                    "Valid core index returned no nested CPUs");
-                result = *(inactive_cpu.begin());
+                result = *(inactive_cpu.rbegin());
+                found_inactive_core = true;
                 break;
             }
         }
+        if (!found_inactive_core) {
+            // If no inactive cores are found, pin the Controller away from the OS.
+            result = num_cpu - 1;
+        }
```

Comment:
To also stay away from the application this if condition can check for free threads:

```
for (int cpu_idx = m_num_cpu - 1; cpu_idx != -1; --cpu_idx) {
    if (!m_is_cpu_active[cpu_idx]) {
        result = cpu_idx;
        break;
    }
}
```

Reply:
issue #1598